Testing.Rmd

```diff
@@ -207,210 +310,434 @@ A test failure looks something like this:
+  "  [4] \"c\" | \"c\" [3]"
+  "  "
+- "       old | new     "
++ "       OLD | NEW     "
+and 3 more ...
+
+* Run `snapshot_accept('diff')` to accept the change
+* Run `snapshot_review('diff')` to interactively review the change
+```
 
-* Global options, like `options()` and `par()`.
+This diff is presented more effectively in most real-world usage, e.g. in the console, by a Git client, or via a Shiny app (see below).
+But even this plain text version highlights the changes quite clearly.
+Each of the two loci of change is indicated with a pair of lines marked with `-` and `+`, showing how the snapshot has changed.
+
+You can call `testthat::snapshot_review('diff')` to review changes locally in a Shiny app, which lets you skip or accept individual snapshots.
+Or, if all changes are intentional and expected, you can go straight to `testthat::snapshot_accept('diff')`.
+Once you've re-synchronized your actual output and the snapshots on file, your tests will pass once again.
+In real life, snapshot tests are a great way to stay informed about changes to your package's user interface, due to your own internal changes or due to changes in your dependencies or even R itself.
+
+`expect_snapshot()` has a few arguments worth knowing about:
+
+* `cran = FALSE`: By default, snapshot tests are skipped if it looks like the
+  tests are running on CRAN's servers.
+  This reflects the typical intent of snapshot tests, which is to proactively
+  monitor user interface, but not to check for correctness, which presumably
+  is the job of other unit tests which are not skipped.
+  In typical usage, a snapshot change is something the developer will want to
+  know about, but it does not signal an actual defect.
+* `error = FALSE`: By default, snapshot code is *not* allowed to throw an error.
+  See `expect_error()`, described above, for one approach to testing errors.
+  But sometimes you want to assess "Does this error message make sense to a
+  human?" and having it laid out in context in a snapshot is a great way to see
+  it with fresh eyes.
+  Specify `error = TRUE` in this case:
+  
+    ```{r eval = FALSE}
+    expect_snapshot(error = TRUE,
+      str_dup(1:2, 1:3)
+    )
+    ```
+  
+* `transform`: Sometimes a snapshot contains volatile, insignificant elements,
+  such as a temporary filepath or a timestamp.
+  The `transform` argument accepts a function, presumably written by you, to
+  remove or replace such changeable text.
+  Another use of `transform` is to scrub sensitive information from the
+  snapshot.
+* `variant`: Sometimes snapshots reflect the ambient conditions, such as the
+  operating system or the version of R or one of your dependencies, and you need
+  a different snapshot for each variant. This is an experimental and somewhat
+  advanced feature, so if you can arrange things to use a single snapshot, you
+  probably should.
+  
+In typical usage, testthat will take care of managing the snapshot files below `tests/testthat/_snaps/`.
```

Comment:
Ok so actually should you document the workflow for when one gets a snapshot failure and needs to fix it? (earlier when explaining snapshot failures)

- I sometimes delete the new one
- I sometimes accept it and reverse the change from the git tab